iceoryx_posh/include/iceoryx_posh/popo/trigger.hpp

```diff
@@ -0,0 +1,122 @@
+
+    template <typename T>
+    using Callback = TriggerState::Callback<T>;
+
+    /// @brief Creates an empty Trigger
+    Trigger() noexcept = default;
+    template <typename T>
+
+    /// @brief Creates a Trigger
+    /// @param[in] origin pointer to the class where the signal originates from, if its set to nullptr the Trigger is in
+    /// a defined but invalid state
+    /// @param[in] conditionVariableDataPtr pointer to the condition variable data, if its set to nullptr the Trigger is
+    /// in a defined but invalid state
+    /// @param[in] hasTriggeredCallback callback to a method which informs the trigger if it was triggered or not. If an
+    /// empty callback is set the trigger is in a defined but invalid state.
+    /// @param[in] resetCallback callback which is called when the trigger goes out of scope.
+    /// @param[in] triggerId id of the trigger
+    /// @param[in] callback function pointer of type void(*)(T * const) to a callback which can be called by the
+    /// trigger.
+    Trigger(T* const origin,
+            const cxx::ConstMethodCallback<bool>& hasTriggeredCallback,
+            const cxx::MethodCallback<void, uint64_t>& resetCallback,
+            const uint64_t triggerId,
+            const Callback<T> callback) noexcept;
+
+    Trigger(const Trigger&) = delete;
+    Trigger& operator=(const Trigger&) = delete;
+    Trigger(Trigger&& rhs) noexcept;
+    Trigger& operator=(Trigger&& rhs) noexcept;
+
+    /// @brief calls reset on destruction
+    ~Trigger();
+
+    /// @brief returns true if the Trigger is valid otherwise false
+    ///        A trigger is valid when:
+    ///         - origin != nullptr
+    ///         - hasTriggeredCallback is set
+    explicit operator bool() const noexcept;
+
+    /// @brief returns true if the trigger is valid otherwise false
+    bool isValid() const noexcept;
+
+    /// @brief returns the result of the provided hasTriggeredCallback
+    bool hasTriggered() const noexcept;
+
+    /// @brief resets and invalidates the Trigger
+    void reset() noexcept;
+
+    /// @brief invalidates the Trigger without calling the reset callback
+    void invalidate() noexcept;
+
+    /// @brief returns the internal unique id of the trigger
+    uint64_t getUniqueId() const noexcept;
+
+    /// @brief returns true if the Triggers are logical equal otherwise false. Two Triggers are logical equal when
+    ///       - origin == rhs.origin
+    ///       - hasTriggeredCallback == rhs.hasTriggeredCallback
+    ///       - triggerId == rhs.triggerId
+    bool isLogicalEqualTo(const Trigger& rhs) const noexcept;
+
+    /// @brief sets a new origin of the trigger
+    /// @param[in] newOrigin pointer to the new origin
+    template <typename T>
+    void updateOrigin(T* const newOrigin) noexcept;
+
+    /// @brief returns the TriggerState
+    TriggerState getTriggerState() const noexcept;
+
+  private:
+    TriggerState m_triggerState;
```

Comment:
Owning all of these makes trigger quite heavy to be handed out in vectors all the time I think. Why not hand out proxy objects that essentially point to a trigger?

Reply:
What if you point to a trigger and a trigger inside the waitset is attached or detached. Then all the triggers would be moved and in a different position and we would have dangling pointers.

Of course we could store the triggers in a list but our main use case is iterating over the triggers and not attach/detaching them. So we lose performance and gain memory efficiency.

I already added this to the issue but I think we need a more sophisticated (and hopefully simple) approach to resolve this problem.